doc/userguide/rules/rules-internals.rst

```diff
@@ -0,0 +1,244 @@
+Rule Processing
+===============
+
+.. toctree::
+
+Suricata rules have many elements that influence how they are processed by
+Suricata and matched against network traffic.
+
+This section explains some key aspects of how Suricata handles rules internally,
+so it can be easier to understand/predict how different rules may interact in
+specific scenarios.
+
+**This material is intended for:** rule writers; developers.
+
+Possible questions one should be better equipped to answer after reading this
+document:
+
+- What happens if two rules have the same priority value (the keyword)?
+- What type of rules will be evaluated first, given a set of rules?
+- How does Suricata decide what rule is "more important" when matching traffic?
+
+.. important:: Rules processing is also heavily affected by rule types, as mentioned
+   in this chapter. You may want to read more on :doc:`rule-types`.
+
+.. note::
+
+    Throughout this documentation and for Suricata, the terms "rule" and
+    "signature" are mainly interchangeable, unless context indicates otherwise.
+
+Overview
+--------
+
+Rules are provided to Suricata via rules files. Starting from those, the
+Detection Engine loader will:
+
+#. Load all the signatures;
+#. Check for validity (non-existing keywords; duplicated ``sid`` etc);
+#. Report stats on loaded, good and bad signatures;
+#. Sort all valid signatures and store them in a list in the Detect Engine
+   Context (``DetectEngineCtx``) structure, taking into consideration several
+   rule aspects according to their order of relevance to the Detection Engine;
+#. Attribute internal rule IDs that reflect this `rule prioritization`_;
+#. During inspection, match rules against the inspected traffic, according to rule
+   and traffic type.
+
+.. _rule prioritization:
+
+Rule Prioritization
+-------------------
+
+Suricata registers several different ordering functions (with
+``SCSigRegisterSignatureOrderingFuncs()``), which are then used to compare
+the rules, sort them, and define their priority. The elements taken into
+consideration for such are the signature's:
+
+#. :ref:`Action <actions>`
+#. Usage of :ref:`flowbits`
```

Comment:
Action and priority are clear orders for me

What is the order about usage of flowbits or other ?
Like if it writes flowbits, it has more priority ? (writes, reads, none) ?
(There is a bug about this @inashivb is working on if a rule reads and write flowbits)

Reply:
Right now, in master, the priority w.r.t. flowbits is: write (highest) -> write + read -> read (lowest).
Yes, this is not always correct but this is how it is supposed to be.

Reply:
And no flowbits is even less than read, right ?